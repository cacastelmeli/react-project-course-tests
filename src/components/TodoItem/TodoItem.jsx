// @ts-check
import React from 'react'

import './TodoItem.scss'

/**
 * @typedef {Object} TodoItem
 * @property {number} id
 * @property {string} text
 * @property {boolean} done
 */

/**
 * @typedef {Object} TodoItemProps
 * @property {TodoItem} item
 * @property {(
 *  item: TodoItem,
 *  checked: boolean
 * ) => void} onItemChecked
 * @property {(item: TodoItem) => void} onItemRemove
 */

/**
 * @type React.FC<TodoItemProps>
 */
const TodoItem = ({
  item,
  onItemChecked,
  onItemRemove
}) => {
  /**
   * @type {React.InputHTMLAttributes<HTMLInputElement>['onChange']}
   */
  const onCheck = ({ target }) => {
    onItemChecked(item, target.checked)
  }

  return (
    <li className="c-todo-item" data-item-id={item.id}>
      <button
        data-testid="removeBtn"
        className="c-todo-item__remove-btn"
        onClick={() => onItemRemove(item)}>
        &times;
      </button>

      <input
        data-testid="toggleCheck"
        className="c-todo-item__done-check"
        type="checkbox"
        onChange={onCheck}
        checked={item.done} />

      <span
        data-testid="text"
        className={`c-todo-item__text ${item.done ? 'c-todo-item__text--done' : ''}`}>
        {item.text}
      </span>
    </li>
  )
}

export default TodoItem
