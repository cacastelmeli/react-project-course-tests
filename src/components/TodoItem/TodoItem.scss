.c-todo-item {
  &__remove-btn {
    
  }

  &__done-check {

  }

  &__text {
    
  }
}