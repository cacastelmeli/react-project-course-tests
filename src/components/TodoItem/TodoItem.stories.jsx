//Â @ts-check
import TodoItem from './TodoItem'

export default {
  title: 'Components/TodoItem',
  component: TodoItem 
}

export const Default = props => (
  <TodoItem
    item={{
      id: Math.random(),
      text: 'Prueba',
      done: false
    }}
    {...props} />
)
