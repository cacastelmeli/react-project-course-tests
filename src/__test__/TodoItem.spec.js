// @ts-check
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import TodoItem from '../components/TodoItem/TodoItem'

/**
 * @type {import('../components/TodoItem/TodoItem').TodoItem}
 */
const testItem = {
  id: 1,
  text: 'test',
  done: false
}

const noop = () => {}

test('should render correctly', () => {
  const { container } = render(
    <TodoItem
      item={testItem}
      onItemChecked={noop}
      onItemRemove={noop} />
  )

  expect(container).toMatchSnapshot()
})

test('should call prop events', async () => {
  let currentTestItem = testItem
  const onItemChecked = jest.fn().mockImplementation((_, done) => (
    currentTestItem = {
      ...currentTestItem,
      done
    }
  ))

  const onItemRemove = jest.fn()

  const renderFn = () => (
    <TodoItem
      item={currentTestItem}
      onItemChecked={onItemChecked}
      onItemRemove={onItemRemove} />
  )

  const { findByTestId, rerender } = render(renderFn())

  const removeBtn = await findByTestId('removeBtn')
  const toggleCheck = await findByTestId('toggleCheck')
  const text = await findByTestId('text')

  fireEvent.click(removeBtn)
  fireEvent.click(toggleCheck)

  // Assert item removing
  expect(onItemRemove).toBeCalledWith(testItem)
  expect(onItemRemove).toBeCalledTimes(1)

  // Assert item's status toggling
  expect(onItemChecked).toBeCalledWith(testItem, true)
  expect(onItemChecked).toBeCalledTimes(1)

  expect(text.classList.contains('c-todo-item__text--done')).toBe(false)

  rerender(renderFn())

  expect(text.classList.contains('c-todo-item__text--done')).toBe(true)
})
