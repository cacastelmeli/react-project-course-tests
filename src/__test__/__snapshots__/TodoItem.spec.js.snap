// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <li
    class="c-todo-item"
    data-item-id="1"
  >
    <button
      class="c-todo-item__remove-btn"
      data-testid="removeBtn"
    >
      Ã—
    </button>
    <input
      class="c-todo-item__done-check"
      data-testid="toggleCheck"
      type="checkbox"
    />
    <span
      class="c-todo-item__text "
      data-testid="text"
    >
      test
    </span>
  </li>
</div>
`;
